src/time-to-interactive/TTIBoomerang.ts

```diff
@@ -0,0 +1,331 @@
+                            clearInterval(tti);
+                            // TTI is Visually Ready + time to start of ok period from visually ready
+                            const timeToQuietPeriodFromVisuallyReady: number =
+                                (currBucket -
+                                    this.REQUIRED_OK_INTERVALS -
+                                    startBucket +
+                                    1) *
+                                this.COLLECTION_PERIOD;
+                            resolve(
+                                visuallyReadyTimestamp +
+                                    timeToQuietPeriodFromVisuallyReady
+                            );
+                            break;
+                        } else {
+                            continue;
+                        }
+                    } else {
+                        okIntervals = 0;
+                    }
+
+                    currBucket = endBucket;
+                }
+                totalIntervals += 1;
+            }, this.CHECK_PERIOD);
+        });
+    }
+
+    initListeners(): void {
+        // record domcontentloaded if its found already
+        if (window.performance) {
+            if (window.performance.getEntriesByType(NAVIGATION)) {
+                this.domContentLoadedEventEnd =
+                    window.performance.getEntriesByType(
+                        NAVIGATION
+                    )[0]?.domContentLoadedEventEnd;
+            }
+        }
+
+        // Use performance observer to record long tasks and domcontentloaded
+        this.eventListener();
+
+        // Record support for FCP and LCP and init listeners
+        this.lcpSupported = isLCPSupported();
+        this.fcpSupported = isFCPSupported();
+
+        // Use web vitals library for LCP and FCP, if supported
+        if (this.fcpSupported) {
+            onFCP((metric) => this.handleWebVitals(metric));
+        }
+        if (this.lcpSupported) {
+            onLCP((metric) => this.handleWebVitals(metric));
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        if (window.requestAnimationFrame(() => {})) {
+            // Init FPS listener if supported
+            this.fpsSupported = true;
+            this.framesPerSecondListener();
+        }
+    }
+
+    private computeTimeWindow(currTime?: number) {
+        if (typeof currTime === 'undefined') {
+            return Math.floor(performance.now() / this.COLLECTION_PERIOD);
+        } else {
+            return Math.floor(currTime / this.COLLECTION_PERIOD);
+        }
+    }
+
+    private addToTracker(type: string, bucket: number, value: any) {
+        if (!this.ttiTracker[type]) {
+            this.ttiTracker[type] = [];
+        }
+
+        if (!this.ttiTracker[type][bucket]) {
+            this.ttiTracker[type][bucket] = 0;
+        }
+
+        this.ttiTracker[type][bucket] += value;
+    }
+
+    eventListener = () => {
```

Comment:
Question: where are we cleaning up this listener?

Reply:
Added logic to stop and disconnect listeners upon TTI resolution or rejection